### PR TITLE
[TORCH] Add support for aten.hinge_embedding_loss Op 

### PR DESCRIPTION
**Pull request by sharavak**
- Decomposed hinge_embedding_loss op into Aten ops.
- Added test cases in the e2e part.

This implementation addresses and closes https://github.com/llvm/torch-mlir/issues/4222

### CODE REVIEW

**Comment by sharavak** [comment]
@stellaraccident , @vivekkhandelwal1, @zjgarvey I’d be grateful if any of you could take a look at this PR. Your feedback would be greatly appreciated!

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/reduction.py [hunk 1]
```diff
@@ -2455,6 +2455,95 @@ def BinaryCrossEntropyWithLogitsStaticModule_basic(module, tu: TestUtils):
...
+            ([1], torch.float32, True),
+        ]
+    )
+    def forward(self, input, target):
+        return torch.ops.aten.hinge_embedding_loss(input, target, margin=1.0)
+
+
+@register_test_case(module_factory=lambda: HingeEmbeddingLossWithoutReductionModule())
+def HingeEmbeddingLossWithoutReductionModule_basic(module, tu: TestUtils):
+    module.forward(tu.rand(8, 5), tu.rand(1))
```
Comment (vivekkhandelwal1) [inline projects/pt1/python/torch_mlir_e2e_test/test_suite/reduction.py]: The default value for the reduction is mean as specified here: https://docs.pytorch.org/docs/stable/generated/torch.nn.HingeEmbeddingLoss.html#torch.nn.HingeEmbeddingLoss:~:text=override%20reduction.-,Default%3A%20%27mean%27,-Shape%3A.

So, this is not a no-reduction case.

Comment (vivekkhandelwal1) [inline projects/pt1/python/torch_mlir_e2e_test/test_suite/reduction.py]: Please use dynamic shapes in the test unless there is a need for static shape tests to be tested or it's the limitation of the lowering.

File: projects/pt1/python/torch_mlir/jit_ir_importer/build_tools/abstract_interp_lib_gen.py [hunk 2]
```diff
@@ -2183,6 +2183,12 @@ def aten〇nll_loss_forward〡shape(self: List[int], target: List[int], weight:
 def aten〇nll_loss_backward〡shape(grad_output: List[int], self: List[int], target: List[int], weight: Optional[List[int]], reduction: int, ignore_index: int, total_weight: List[int]) -> List[int]:
     return upstream_shape_functions.unary(self)
 
+def aten〇hinge_embedding_loss〡shape(self: List[int], target: List[int], margin: float = 1., reduction: int = 1) -> List[int]:
+    if reduction in [1,2]:
+        return []
+    else:
```
Comment (vivekkhandelwal1) [inline projects/pt1/python/torch_mlir/jit_ir_importer/build_tools/abstract_interp_lib_gen.py]: Else is not required.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 3]
```diff
@@ -10471,6 +10471,89 @@ class DecomposeAtenNllLossForwardOp
...
+        loc, inputTy, targetNotOne, marginClamp, zero);
+    // Compute mask: target != -1
+    auto targetNotMinusOne =
+        rewriter.create<AtenNeScalarOp>(loc, boolType, target, minusOne);
+    // If target != -1 use the original input. Otherwise 0.
+    auto outputSelf = rewriter.create<AtenWhereScalarOtherOp>(
+        loc, inputTy, targetNotMinusOne, input, zero);
+    // Add : outputMargin + outputSelf
+    auto output = rewriter.create<AtenAddTensorOp>(loc, inputTy, outputMargin,
+                                                   outputSelf, /*alpha=*/alpha);
```
Comment (vivekkhandelwal1) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:10626]: Instead of doing all this, you can just do:
```
auto result = rewriter.create<AtenWhereScalarOtherOp>(
        loc, inputTy, targetNotOne, marginClamp, input);
```

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 4]
```diff
@@ -10471,6 +10471,89 @@ class DecomposeAtenNllLossForwardOp
...
+    int64_t reduction;
+    if (!matchPattern(op.getReduction(), m_TorchConstantInt(&reduction))) {
+      return rewriter.notifyMatchFailure(op,
+                                         "reduction should be a constant int!");
+    }
```
Comment (vivekkhandelwal1) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp]: Place this code above so that if this condition does not hold true the code can exit early.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 5]
```diff
@@ -10471,6 +10471,89 @@ class DecomposeAtenNllLossForwardOp
...
+    // reduction: mean
+    if (reduction == 1) {
+      loss = rewriter.create<AtenMeanOp>(loc, resultTy, output, none);
+    } else if (reduction == 2) {
+      // reduction: sum
+      loss = rewriter.create<AtenSumOp>(loc, resultTy, output, none);
+    } else {
+      // reduction: none
+      loss = output;
+    }
```
Comment (vivekkhandelwal1) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp]: Instead of this, do:
```
Value loss = output;
```

**Review by vivekkhandelwal1 — CHANGES_REQUESTED** [review]

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp (hunk 3 above)
Comment (sharavak) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:10626]: @vivekkhandelwal1 Thanks for the suggestion, I did try the simplified version initially, but it caused numerical validation errors in some test cases. This happens because the target tensor can sometimes have values other than just -1 and 1.

To handle this properly and stay consistent with PyTorch's semantics, I decided to explicitly check for both target == 1 and target == -1. This way, the behavior stays correct even if target have values other than just -1 and 1.

Eg:
```
import torch
input=torch.randn(2,3)
target=torch.randn(2,3)
torch.hinge_embedding_loss(input,target)

Output:
tensor([[1.1361, 1.0000, 1.0000],
        [1.4880, 1.1624, 1.0000]])
```

**Comment by sharavak** [comment]
@vivekkhandelwal1 Thanks a lot for the feedback. I've updated the code.

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/reduction.py (hunk 1 above)
Comment (sharavak) [reply projects/pt1/python/torch_mlir_e2e_test/test_suite/reduction.py]: Yes that's correct. Sorry for the wrong naming for that test case. It should be HingeEmbeddingLossReductionNoneModule.

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/reduction.py [hunk 6]
```diff
@@ -2455,6 +2455,95 @@ def BinaryCrossEntropyWithLogitsStaticModule_basic(module, tu: TestUtils):
...
+            ([-1], torch.float32, True),
+        ]
+    )
+    def forward(self, input, target):
+        return torch.ops.aten.hinge_embedding_loss(input, target, margin=1.0)
+
+
+@register_test_case(module_factory=lambda: HingeEmbeddingLossReductionNoneModule())
+def HingeEmbeddingLossReductionNoneModule_basic(module, tu: TestUtils):
+    module.forward(tu.rand(8, 5), tu.rand(1))
```
Comment (vivekkhandelwal1) [inline projects/pt1/python/torch_mlir_e2e_test/test_suite/reduction.py]: All 3 of these tests have only a single value as a target; as a result, not all the paths of the lowering will get tested. Ideally, for testing purposes, the target tensor should contain a mix of `-1` and `1` values. Also, how do you make sure that the target tensor contains only valid values?

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp (hunk 3 above)
Comment (vivekkhandelwal1) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:10626]: > This happens because the target tensor can sometimes have values other than just -1 and 1.

In what cases and how? Since the definition says that it can contain only `-1` and `1`.

Comment (sharavak) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:10626]: Thanks for the reply @vivekkhandelwal1, I got reference from the  Pytorch native implementation [https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/Loss.cpp#L182](url). While the official definition of hinge_embedding_loss states that the target should contain only -1 and 1, the native implementation doesn’t enforce this restriction and handles arbitrary values using at::where(target != 1, ) and at::where(target != -1,). So to stay with the Pytorch behaviour, I followed the same logic.

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/reduction.py (hunk 6 above)
Comment (sharavak) [reply projects/pt1/python/torch_mlir_e2e_test/test_suite/reduction.py]: Thanks for the review @vivekkhandelwal1 
I will update the target tensor.

>  Also, how do you make sure that the target tensor contains only valid values?

For this question, I’ve addressed that in my earlier comment here: [#4227 (comment)](https://github.com/llvm/torch-mlir/pull/4227#discussion_r2176421042). @vivekkhandelwal1.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp (hunk 3 above)
Comment (vivekkhandelwal1) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:10626]: Okay. In that case, it would be better to add this link and the justification in the comment. So that, the future users/contributors may not get confused.

Comment (sharavak) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:10626]: Thanks for the reply @vivekkhandelwal1. I have added the comments as per your suggestion.
